Use Java 17 features

Gold856 opened the pull request:
Uses enhanced instanceof (and simplify `equals` methods)
Uses switch expressions and arrow labels
Seal and finalize some Shuffleboard classes

Starlight220 commented:
> apparently none of the linters can handle an exhaustive switch expression with no default

Are you handling the case of `null`? (Or otherwise proving it's not null)

Starlight220 commented on cscore/src/main/java/edu/wpi/first/cscore/CvSink.java:
```diff
@@ -32,27 +32,12 @@ public void close() {
-      case kBGRA:
-        type = CvType.CV_8UC4;
-        break;
-      case kGray:
-      case kMJPEG:
```
I'm not a fan of these two being inlined implicitly into the default case. They should be included explicitly as in the old impl.

Gold856 commented:
> Are you handling the case of `null`? (Or otherwise proving it's not null)

That's not available until Java 21. Switch expressions/statements have historically thrown NPEs until Java 21. And some of the tools say they won't warn on switch expressions because javac will check switch exhaustiveness, but they still warn anyways.

Gold856 commented:
Apparently switch expressions are separate from arrow labels, so when the linters were warning about missing switch defaults on switch statements, they were, in fact, warning about missing switch defaults on switch statements. Defaults have been put back in, matching whatever previous logic was present.

sciencewhiz commented:
Did you do this by hand, or did you use a tool?

Gold856 commented:
I cherry-picked @SamCarlberg's commit that refactored a bunch of code to use Java 21 features, removed the Java 21 parts to make it Java 17 compatible, and refactored any remaining code, mostly by using VS Code's search feature with the regexes `case .*:` to convert case labels, `else if` to search for places where a switch would be better, and `instanceof` to convert instanceof usages.

Gold856 commented on cscore/java-examples/RawCVMatSink.java:
```diff
@@ -20,22 +20,12 @@ public class RawCVMatSink extends ImageSink {
-        break;
-    }
-    return type;
+    return switch (pixelFormat) {
+      case kYUYV, kRGB565, kY16, kUYVY -> CvType.CV_8UC2;
+      case kBGR -> CvType.CV_8UC3;
+      case kBGRA -> CvType.CV_8UC4;
+      case kGray, kMJPEG, kUnknown -> CvType.CV_8UC1;
+    };
```
Yeah, I assumed that we want this to match `CvSink`.
